### PR TITLE
Fix some errors reported via ASAN / UBSAN

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fixes a few errors reported by address sanitizers / memory sanitizers
 - Adjust a few config flags specific to RelWithDebInfo
 - Increase DISCOVERY timeout for gtest, I get bit by that from time to time and in Debug with ASAN + UBSAN I 100% get the error since it's much slower

Note: a RelWithDebInfo build on Ubuntu 20.04 / GCC 10 with ASAN and UBSAN enabled is SUPPPPER Slow. Like more than 2 hours on my 8 cores / 16 threads machine that builds under 30 minutes. Same build on M1 Mac with clang 14 was much faster too. I think it's due to `var-tracking-assignments` in GCC which is enabled if both DEBUG symbols and optimizations are enabled simultaneously. I do get some limits exceeded: `variable tracking size limit exceeded with -fvar-tracking-assignments`

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -357,12 +357,18 @@ if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" OR CMAKE_CXX_COMPILER_ID MATCHES ".*Clan
     set(SANITIZER_LINKER_FLAGS "${SANITIZER_LINKER_FLAGS} -fsanitize=undefined")
   endif()
 
+  option(ENABLE_NORECOVER_SANITIZER  "When any sanitizer that supports it report an error, exit with non zero" OFF)
+  if(ENABLE_NORECOVER_SANITIZER)
+    add_definitions(-fno-sanitize-recover=all)
+    set(SANITIZER_LINKER_FLAGS "${SANITIZER_LINKER_FLAGS} -fno-sanitize-recover=all")
+  endif()
```
Comment (jmarrec) [inline CMakeLists.txt:364]: new advanced option, turned off by default.

File: CMakeLists.txt [hunk 2]
```diff
@@ -515,6 +521,9 @@ if(UNIX)
     if (NOT CMAKE_CXX_COMPILER_ID MATCHES "Clang")
       # DLM: had to add this due to issues with boost optional and gcc, may be resolved if we move to std::optional
       set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-maybe-uninitialized")
+      if (CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo")
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-uninitialized")
+      endif()
```
Comment (jmarrec) [inline CMakeLists.txt:526]: Needed that on Ubuntu with GCC 10. It thinks that `return boost::none` is uninitialized.

File: ProjectMacros.cmake [hunk 3]
```diff
@@ -34,7 +34,8 @@ macro(CREATE_TEST_TARGETS BASE_NAME SRC DEPENDENCIES)
 
     # Tell cmake to discover tests by calling test_exe --gtest_list_tests
     gtest_discover_tests(${BASE_NAME}_tests
-      PROPERTIES TIMEOUT 660
+      PROPERTIES TIMEOUT 660 # Test execution
+      DISCOVERY_TIMEOUT 60   # Time to wait for the test to enumerate available tests (default is 5s, which can fail for us especially in Debug with Sanitizers)
```
Comment (jmarrec) [inline ProjectMacros.cmake:38]: Increase the **Discovery** timeout for the Gtest.

File: src/model/test/AirflowNetworkExternalNode_GTest.cpp [hunk 4]
```diff
@@ -45,12 +45,10 @@ TEST_F(ModelFixture, AirflowNetwork_ExternalNode) {
...
-  Curve curve0 = extnode0.windPressureCoefficientCurve();
-  auto* cpt = static_cast<CurveLinear*>(&curve0);
-  //CurveLinear *cpt = dynamic_cast<CurveLinear*>(&curve0);
-  //ASSERT_NE(nullptr, cpt);
-  EXPECT_EQ(1.0, cpt->coefficient1Constant());
-  EXPECT_EQ(0.0, cpt->coefficient2x());
+  auto cpt_ = extnode0.windPressureCoefficientCurve().optionalCast<CurveLinear>();
+  ASSERT_TRUE(cpt_);
+  EXPECT_EQ(1.0, cpt_->coefficient1Constant());
+  EXPECT_EQ(0.0, cpt_->coefficient2x());
```
Comment (jmarrec) [inline src/model/test/AirflowNetworkExternalNode_GTest.cpp:51]: Fix memory leak in AirflowNetwork gtest

File: src/utilities/floorspace/FSModel.cpp [hunk 5]
```diff
@@ -1050,13 +1050,11 @@ boost::optional<FSVertex> FSDaylightingControl::vertex() const {
...
-const Point3d& FSDaylightingControl::point() const {
+Point3d FSDaylightingControl::point() const {
   if (m_vertex.has_value() && m_definition.has_value()) {
-    return *new Point3d(m_vertex->x(), m_vertex->y(), m_definition->height());
-  } else {
-
-    return *new Point3d(0, 0, 0);
+    return {m_vertex->x(), m_vertex->y(), m_definition->height()};
   }
+  return {0.0, 0.0, 0.0};
```
Comment (jmarrec) [inline src/utilities/floorspace/FSModel.cpp:1057]: new without a delete

File: src/utilities/floorspace/FSModel.hpp [hunk 6]
```diff
@@ -249,8 +249,8 @@ class UTILITIES_API FSDaylightingControl : public FSBase
   void Accept(FSVisitor& visitor) const override;
 
   boost::optional<FSVertex> vertex() const;
-  const Point3d& point() const;
   boost::optional<FSDaylightingControlDefinition> definition() const;
+  Point3d point() const;
```
Comment (jmarrec) [inline src/utilities/floorspace/FSModel.hpp:253]: Changed return type. Won't affect anyone, this is an internal object

**Review by tijcolem — APPROVED** [review]

**Pull request merged**